zebrad/src/commands/start/sync.rs

```diff
@@ -14,37 +14,32 @@ use zebra_consensus::checkpoint;
 use zebra_network::{self as zn, RetryLimit};
 use zebra_state::{self as zs};
 
-pub struct Syncer<ZN, ZS, ZV>
-where
-    ZN: Service<zn::Request>,
-{
-    pub peer_set: ZN,
-    pub state: ZS,
-    pub verifier: ZV,
-    pub retry_peer_set: Retry<RetryLimit, ZN>,
-    pub prospective_tips: HashSet<BlockHeaderHash>,
-    pub block_requests: FuturesUnordered<ZN::Future>,
-    pub fanout: NumReq,
-}
+// XXX in the future, we may not be able to access the checkpoint module.
+const FANOUT: usize = checkpoint::MAX_QUEUED_BLOCKS_PER_HEIGHT;
+/// Controls how far ahead of the chain tip the syncer tries to download before
+/// waiting for queued verifications to complete. Set to twice the maximum
+/// checkpoint distance.
+const LOOKAHEAD_LIMIT: usize = 2 * 2_000;
 
-impl<ZN, ZS, ZC> Syncer<ZN, ZS, ZC>
+#[derive(Debug)]
+pub struct Syncer<ZN, ZS, ZV>
 where
-    ZN: Service<zn::Request> + Clone,
+    ZN: Service<zn::Request, Response = zn::Response, Error = Error> + Send + Clone + 'static,
+    ZN::Future: Send,
+    ZS: Service<zs::Request, Response = zs::Response, Error = Error> + Send + Clone + 'static,
+    ZS::Future: Send,
+    ZV: Service<Arc<Block>, Response = BlockHeaderHash, Error = Error> + Send + Clone + 'static,
+    ZV::Future: Send,
 {
-    pub fn new(peer_set: ZN, state: ZS, verifier: ZC) -> Self {
-        let retry_peer_set = Retry::new(RetryLimit::new(3), peer_set.clone());
-        Self {
-            peer_set,
-            state,
-            verifier,
-            retry_peer_set,
-            block_requests: FuturesUnordered::new(),
-            // Limit the fanout to the number of chains that the
-            // CheckpointVerifier can handle
-            fanout: checkpoint::MAX_QUEUED_BLOCKS_PER_HEIGHT,
-            prospective_tips: HashSet::new(),
-        }
-    }
+    /// Used to perform extendtips requests, with no retry logic (failover is handled using fanout).
+    tip_network: ZN,
+    /// Used to download blocks, with retry logic.
+    block_network: Retry<RetryLimit, ZN>,
+    state: ZS,
+    verifier: ZV,
+    prospective_tips: HashSet<BlockHeaderHash>,
+    pending_blocks:
+        Pin<Box<FuturesUnordered<Instrumented<JoinHandle<Result<BlockHeaderHash, Error>>>>>>,
```

Comment:
type alias might help here

Reply:
I would add one if this was used in multiple places but I didn't add one here because we don't ever need to refer to this type except in this struct definition.

Reply:
fair, I was mostly concerned about the weird wrapping because the type is so long.